src/Dscp/Educator/Web/Student/Queries.hs

```diff
@@ -0,0 +1,351 @@
+{-# LANGUAGE DataKinds                 #-}
+{-# LANGUAGE ExistentialQuantification #-}
+{-# LANGUAGE QuasiQuotes               #-}
+{-# LANGUAGE TypeOperators             #-}
+
+module Dscp.Educator.Web.Student.Queries
+    ( module Dscp.Educator.Web.Student.Queries
+    ) where
+
+import Control.Lens (from, mapping)
+import Data.Coerce (coerce)
+import Data.Time.Clock (UTCTime)
+import Database.SQLite.Simple ((:.) (..), Only (..), Query, ToRow (..))
+import Database.SQLite.Simple.ToField (ToField)
+import Loot.Log (MonadLogging)
+import Text.InterpolatedString.Perl6 (q)
+
+import Dscp.Core.Serialise ()
+import qualified Dscp.Core.Types as Core
+import Dscp.Crypto (Hash)
+import Dscp.DB.SQLite (DomainError (..), MonadSQLiteDB (..), TxBlockIdx (TxInMempool))
+import qualified Dscp.DB.SQLite.Queries as Base
+import Dscp.Educator.Txs (PrivateTx)
+import Dscp.Util (Id, assertJust, listToMaybeWarn)
+import Dscp.Util.Aeson (AsByteString (..))
+
+import Dscp.Educator.Serialise ()
+import Dscp.Educator.Web.Student.Error (APIError (..))
+import Dscp.Educator.Web.Student.Types (Assignment (..), BlkProof (..), Course (..), Grade (..),
+                                        IsEnrolled (..), IsFinal (..), Student, Submission (..))
+import Dscp.Educator.Web.Student.Util (assignTypeRaw)
+
+type MonadStudentAPIQuery m =
+    ( MonadSQLiteDB m
+    , MonadLogging m
+    , MonadThrow m
+    )
+
+getCourse
+    :: MonadStudentAPIQuery m
+    => Id Student -> Core.Course -> m Course
+getCourse studentId courseId =
+    transaction $ do
+        mcourse <-
+            query queryText (Only courseId)
+            >>= listToMaybeWarn "courses"
+        Only desc <- pure mcourse `assertJust` CourseDoesNotExist courseId
+
+        cIsEnrolled <- Base.isEnrolledTo studentId courseId
+        cSubjects <- Base.getCourseSubjects courseId
+        return Course{ cId = courseId, cDesc = desc, .. }
+  where
+    queryText :: Query
+    queryText = [q|
+        select    desc
+        from      Courses
+        where     id = ?
+    |]
+
+getCourses
+    :: MonadStudentAPIQuery m
+    => Id Student -> Maybe IsEnrolled -> m [Course]
+getCourses studentId (coerce -> isEnrolledF) = do
+    let enrolledFilter = maybe "" enrolledClause isEnrolledF
+
+    transaction $ do
+        courses <- query (queryText <> enrolledFilter) (Only studentId)
+
+        forM courses $ \(courceId, desc) -> do
+            cIsEnrolled <- Base.isEnrolledTo studentId courceId
+            cSubjects <- Base.getCourseSubjects courceId
```

Comment:
Wow, `n` queries to database, how cool is that! (it is not, it could be done with at least constant number of queries, but I suppose we shouldn't bother too much about it yet until benchmarks show that this is prohibitively slow)